scg-system/src/main/java/io/telicent/backup/services/DatasetBackupService.java

```diff
@@ -0,0 +1,276 @@
+        response.put("date", DateTimeUtils.nowAsString("yyyy-MM-dd_HH-mm-ss"));
+        response.set("backups", populateNodeFromDirNumerically(getBackUpDir()));
+        return response;
+    }
+
+    public ObjectNode restoreDatasets(String restoreId) {
+        ObjectNode response = MAPPER.createObjectNode();
+        String restorePath = getBackUpDir() + "/" + restoreId;
+        response.put("restorePath", restorePath);
+        if (!checkPathExistsAndIsDir(restorePath)) {
+            response.put("reason", "Restore path unsuitable: " + restorePath);
+            response.put("success", false);
+        } else {
+            List<String> datasets = getSubdirectoryNames(restorePath);
+            if (datasets.isEmpty()) {
+                response.put("reason", "Restore path unsuitable: " + restorePath);
+                response.put("success", false);
+            } else {
+                for (String datasetName : datasets) {
+                    response.set(datasetName, restoreDataset(restorePath, datasetName));
+                }
+            }
+        }
+        return response;
+    }
+
+    ObjectNode restoreDataset(String restorePath, String datasetName) {
+        ObjectNode response = MAPPER.createObjectNode();
+        response.put("dataset-id", datasetName);
+        DataAccessPoint dataAccessPoint = dapRegistry.get("/" + datasetName);
+        if (dataAccessPoint == null) {
+            response.put("reason", datasetName + " does not exist");
+            response.put("success", false);
+            return response;
+        }
+        DatasetGraph dsg = dataAccessPoint.getDataService().getDataset();
+        response.set("tdb", restoreTDB(datasetName, dsg, restorePath));
+        response.set("labels", restoreLabelStore(datasetName, dsg, restorePath));
+        return response;
+    }
+
+    ObjectNode restoreTDB(String datasetName, DatasetGraph dsg, String restorePath) {
+        ObjectNode node = MAPPER.createObjectNode();
+        String tdbRestoreFile = restorePath + "/" + datasetName + "/TDB/" + datasetName + "_backup.nq.gz";
+        node.put("restorePath", tdbRestoreFile);
+        if (!checkPathExistsAndIsFile(tdbRestoreFile)) {
+            node.put("reason", "Restore file not found: " + tdbRestoreFile);
+            node.put("success", false);
+            return node;
+        }
+        try {
+            executeRestoreTDB(dsg, tdbRestoreFile);
+            node.put("success", true);
+        } catch (Exception e) {
+            node.put("reason", e.getMessage());
+            node.put("success", false);
+        }
+        return node;
+    }
+
+    @ExcludeFromJacocoGeneratedReport
+    void executeRestoreTDB(DatasetGraph dsg, String tdbRestoreFile) throws IOException {
+        try (InputStream fis = new FileInputStream(tdbRestoreFile);
+             InputStream gis = new GZIPInputStream(fis)) {
+            Txn.executeWrite(dsg, () -> {
+                dsg.clear();
```

Comment:
Again for future ticket, may want to see if there's other ways to achieve this.  `dsg.clear()` is one of those operations which feels like it should be cheap but can be horrendously expensive on large datasets.  It might almost be better to do something akin to how TDB's compact operation works whereby you create the next `Data_NNN` directory, use that as a blank database and load the backup into that, then switch the TDB instance to point to the new directory.  That's getting fairly into the low level weeds of TDB so not something to worry about immediately.